clippy_lints/src/methods/unnecessary_min_max.rs

```diff
@@ -0,0 +1,103 @@
+use clippy_utils::source::snippet;
+use clippy_utils::{clip, int_bits, unsext};
+use hir::Expr;
+
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+
+use rustc_middle::ty;
+use rustc_span::Span;
+
+pub fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'_>,
+    name: &str,
+    recv: &'tcx Expr<'_>,
+    arg: &'tcx Expr<'_>,
+) {
+    let typeck_results = cx.typeck_results();
+    if let (Some(left), Some(right)) = (constant(cx, typeck_results, recv), constant(cx, typeck_results, arg)) {
+        let Some(ord) = Constant::partial_cmp(cx.tcx, typeck_results.expr_ty(recv), &left, &right) else {
+            return;
+        };
+
+        lint(cx, expr, name, recv.span, arg.span, ord);
+    }
+
+    if let Some(extrema) = detect_extrema(cx, recv) {
+        let ord = match extrema {
+            Extrema::Minimum => Ordering::Less,
+            Extrema::Maximum => Ordering::Greater,
+        };
+        lint(cx, expr, name, recv.span, arg.span, ord);
+    }
+    if let Some(extrema) = detect_extrema(cx, arg) {
+        let ord = match extrema {
+            Extrema::Minimum => Ordering::Greater,
+            Extrema::Maximum => Ordering::Less,
+        };
+        lint(cx, expr, name, recv.span, arg.span, ord);
+    }
+}
+
+fn lint(cx: &LateContext<'_>, expr: &Expr<'_>, name: &str, lhs: Span, rhs: Span, order: Ordering) {
+    let cmp_str = if (name == "min" && order.is_ge()) || (name == "max" && order.is_le()) {
+        "smaller"
+    } else {
+        "greater"
+    };
+
+    let suggested_value = if (name == "min" && order.is_ge()) || (name == "max" && order.is_le()) {
+        snippet(cx, rhs, "..")
+    } else {
+        snippet(cx, lhs, "..")
+    };
+
+    let msg = format!(
+        "`{}` is never {} than `{}` and has therefore no effect",
+        snippet(cx, lhs, ".."),
+        cmp_str,
+        snippet(cx, rhs, "..")
+    );
+    span_lint_and_sugg(
+        cx,
+        UNNECESSARY_MIN_MAX,
+        expr.span,
+        &msg,
+        "try",
+        suggested_value.to_string(),
+        Applicability::MachineApplicable,
+    );
+}
+
+#[derive(Debug)]
+enum Extrema {
+    Minimum,
+    Maximum,
+}
+fn detect_extrema<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<Extrema> {
+    let ty = cx.typeck_results().expr_ty(expr);
+
+    let cv = constant(cx, cx.typeck_results(), expr)?;
+
+    match (ty.kind(), cv) {
+        (&ty::Uint(_), Constant::Int(0)) => Some(Extrema::Minimum),
+        (&ty::Int(ity), Constant::Int(i)) if i == unsext(cx.tcx, i128::MIN >> (128 - int_bits(cx.tcx, ity)), ity) => {
+            Some(Extrema::Minimum)
+        },
+
+        (&ty::Int(ity), Constant::Int(i)) if i == unsext(cx.tcx, i128::MAX >> (128 - int_bits(cx.tcx, ity)), ity) => {
+            Some(Extrema::Maximum)
+        },
+        (&ty::Uint(uty), Constant::Int(i)) if i == clip(cx.tcx, u128::MAX, uty) => Some(Extrema::Maximum),
+
+        _ => None,
+    }
```

Comment:
`unsext`/`clip`/`int_bits` handle `usize`/`isize` but for this lint we wouldn't want that (excluding `0usize`) so that `x.min(i32::MAX as isize)` doesn't lint on 32 bit targets

`cv.int_value(...)?` could replace the `unsext`s with `ity/uty.bit_width()` instead of `int_bits` to exclude `*size`

Reply:
> ensure x.min(i32::MAX as isize) doesn't lint on 32 bit targets
I don't really under stand this. If x is of i32, it shouldn't be able to be compared to `i32::MAX as isize`?

Reply:
In that example it would be that `x` is an `isize`, it wasn't a reference to the existing `i32` `x` in the test file